src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -250,7 +285,309 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+                    Criticality::NonCritical,
+                    None,
+                ),
+                // 5280 4.2.1.3: Key Usage
+                ExtensionPolicy::maybe_present(KEY_USAGE_OID, Criticality::Agnostic, None),
+                // CA/B 7.1.2.7.12 Subscriber Certificate Subject Alternative Name
+                ExtensionPolicy::present(
+                    SUBJECT_ALTERNATIVE_NAME_OID,
+                    Criticality::Agnostic,
+                    Some(ee::subject_alternative_name),
+                ),
+                // 5280 4.2.1.9: Basic Constraints
+                ExtensionPolicy::maybe_present(
+                    BASIC_CONSTRAINTS_OID,
+                    Criticality::Agnostic,
+                    Some(ee::basic_constraints),
+                ),
+                // 5280 4.2.1.10: Name Constraints
+                ExtensionPolicy::not_present(NAME_CONSTRAINTS_OID),
+                // CA/B 7.1.2.7.10 Subscriber Certificate Extended Key Usage
+                ExtensionPolicy::present(
+                    EXTENDED_KEY_USAGE_OID,
+                    Criticality::NonCritical,
+                    Some(ee::extended_key_usage),
+                ),
+            ]),
+        }
+    }
+
+    fn permits_basic(&self, cert: &Certificate<'_>) -> Result<(), PolicyError> {
+        let extensions = cert.extensions()?;
+
+        // CA/B 7.1.1:
+        // Certificates MUST be of type X.509 v3.
+        if cert.tbs_cert.version != 2 {
+            return Err("certificate must be an X509v3 certificate".into());
+        }
+
+        // 5280 4.1.1.2 / 4.1.2.3: signatureAlgorithm / TBS Certificate Signature
+        // The top-level signatureAlgorithm and TBSCert signature algorithm
+        // MUST match.
+        if cert.signature_alg != cert.tbs_cert.signature_alg {
+            return Err("mismatch between signatureAlgorithm and SPKI algorithm".into());
+        }
+
+        // 5280 4.1.2.2: Serial Number
+        let serial_bytes = cert.tbs_cert.serial.as_bytes();
+        if serial_bytes.len() == 1 && serial_bytes[0] == 0 {
+            // The serial number MUST be a positive integer.
+            return Err("certificate serial number must not be 0".into());
+        } else if !(1..=21).contains(&serial_bytes.len()) {
+            // Conforming CAs MUST NOT use serial numbers longer than 20 octets.
+            // NOTE: In practice, this requires us to check for an encoding of
+            // 21 octets, since some CAs generate 20 bytes of randomness and
+            // then forget to check whether that number would be negative, resulting
+            // in a 21-byte encoding.
+            return Err("certificate must have a serial between 1 and 20 octets".into());
+        } else if serial_bytes[0] & 0x80 == 0x80 {
+            // If the high-bit is set, then we know the serial number is negative.
+            return Err("certificate serial number cannot be negative".into());
+        }
```

Comment:
CC @alex how do you feel about this versus an `is_negative` helper?

Reply:
Maybe send a PR to rust-asn1 adding an `is_negative` method :-) + a TODO to use it here. But this is fine for now.

Reply:
@woodruffw @alex Seems like we can't trigger this through Limbo since we check for negative values [here](https://github.com/pyca/cryptography/blob/main/src/rust/src/x509/certificate.rs#L871).

I wonder whether it's worth testing it via the Rust API or should we remove the check entirely and leave a comment explaining why.

Reply:
@woodruffw @alex Seems like we can't trigger this through Limbo since we're checking for negative values [here](https://github.com/pyca/cryptography/blob/main/src/rust/src/x509/certificate.rs#L871). It might be easier to test this via Rust.

Reply:
Works for me either way! In terms of doing it from Python: there might be a testcase in `cryptography_vectors` that we can reuse.

Reply:
Opened https://github.com/alex/rust-asn1/pull/425 for the upstream change.